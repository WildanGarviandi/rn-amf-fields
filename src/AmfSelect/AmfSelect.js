import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, Text, View, Picker } from 'react-native'
import Icon from 'react-native-vector-icons/MaterialIcons'
import { Dropdown } from 'react-native-material-dropdown'

class AmfSelect extends Component {
  constructor(props) {

    super(props)

    this.state = {
      valid: false,
      isFocused: false
    }

  }

  componentDidMount() {
    this.validate()
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.isFocused !== this.state.isFocused) {
      if (!this.state.isFocused) this.validate()
    }
  }

  getLabel = (val) => {
    let option = this.props.options.find(i => i.value === val)
    return option && option.label
  }

  inputValid = () => {
    const { validation, label } = this.props
    if (!validation) return ({status: true})

    // check if pristine
    if (this.props.value === null || this.props.value === undefined) return ({status: true})

    // check if required
    if (validation.required) {
      if (this.props.value === '') return ({status: false, message: label + ' harus dipilih'})
    }

    return ({status: true})
  }

  onSelectWillShow = () => {
    if (this.props.validation && this.props.validation.required && this.props.value === null) {
      this.props.onChange('')
    }

    this.setState({isFocused: true})
  }

  onSelectWillHide = () => {
    this.setState({isFocused: false})
  }

  validate = () => {
    let validationObject = this.inputValid()
    this.setState({valid: validationObject.status, errorMessage: validationObject.message})
  }

  render() {
    const { layout, value, label, options, onChange } = this.props

    let validationStyle = {}
    let validationText
    
    if (!this.state.valid) {
      validationStyle = {borderColor: 'red', borderWidth: 1}
      validationText = <Text style={style.errorText}>{this.state.errorMessage}</Text>
    }

    let labelText
    if (this.props.label) {
      labelText = <Text style={[style.label, {marginBottom: layout == 'vertical' ? 8 : 0}]}>{this.props.label}</Text>
    }

    return (
      <View>
        <Dropdown
          label={label}
          data={options}
			    value={value || ''}
          onChangeText={onChange}
          error=""
			  />
        { validationText }
      </View>
    )
  }
}

const style = StyleSheet.create({
  container: {},
  containerHorizontal: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  dropdownContainer: {
    elevation: 4,
    borderRadius: 25,
    borderWidth: 1,
    borderColor: 'transparent',
    backgroundColor: 'white',
    justifyContent: 'center'
  },
  dropdown: {
    width: 100,
    elevation: 4,
    padding: 8,
    height: 150
  },
  errorText: {
    color: 'red',
    margin: 4,
    marginBottom: 0
  },
  label: {
    color: '#454545',
    marginRight: 16
  },
  option: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: 30
  },
  icon: {
    marginRight: 16
  },
  valueText: {
    marginLeft: 16
  }
})

const itemShape = PropTypes.shape({
  label: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired
})

AmfSelect.defaultProps = {
  layout: 'vertical'
}

AmfSelect.propTypes = {
  disabled: PropTypes.bool,
  dropdownStyle: PropTypes.object,
  options: PropTypes.arrayOf(itemShape).isRequired,
  label: PropTypes.string,
  layout: PropTypes.oneOf(['vertical', 'horizontal']),
  onChange: PropTypes.func.isRequired,
  optionStyle: PropTypes.object,
  selector: PropTypes.element,
  style: PropTypes.object,
  validation: PropTypes.object,
  value: PropTypes.any,
}

export default AmfSelect
