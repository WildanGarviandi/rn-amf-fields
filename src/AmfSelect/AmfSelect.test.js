import React from 'react'
import { Text } from 'react-native'
import AmfSelect from './AmfSelect'
import renderer from 'react-test-renderer'

let options = [{
  label: 'foo',
  value: 'foo'
}, {
  label: 'bar',
  value: 'bar'
}]
let onChange = jest.fn()
let validation = {
  required: true
}

test('AmfSelect without validation', () => {
  let props = { options, onChange }

  const component = renderer.create(<AmfSelect {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfSelect with label', () => {
  let props = { options, onChange, label: 'Label' }
  let propsHorizontal = { ...props, layout: 'horizontal' }

  const component = renderer.create(<AmfSelect {...props} />)
  const componentHorizontal = renderer.create(<AmfSelect {...propsHorizontal} />)
  
  let tree = component.toJSON()
  let treeHorizontal = componentHorizontal.toJSON()
  expect(tree).toMatchSnapshot()
  expect(treeHorizontal).toMatchSnapshot()
})

test('AmfSelect with custom selector', () => {
  let props = {
    options, onChange,
    selector: <Text>Custom Selector</Text>
  }

  const component = renderer.create(<AmfSelect {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfSelect with horizontal layout', () => {
  let props = {
    options, onChange,
    layout: 'horizontal'
  }

  const component = renderer.create(<AmfSelect {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfSelect with required validation', () => {
  let props = {
    options, onChange, validation
  }
  let propsWithValue = {
    ...props, value: 'foo'
  }

  const component = renderer.create(<AmfSelect {...props} />)
  const componentWithValue = renderer.create(<AmfSelect {...propsWithValue} />)
  
  let tree = component.toJSON()
  let treeWithValue = componentWithValue.toJSON()
  expect(tree).toMatchSnapshot()
  expect(treeWithValue).toMatchSnapshot()
})

test('AmfSelect with unmatch value', () => {
  let props = {
    options, onChange, validation,
    value: 'No Match'
  }

  const component = renderer.create(<AmfSelect {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfSelect with empty value', () => {
  let props = {
    options, onChange, validation,
    value: ''
  }

  const component = renderer.create(<AmfSelect {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfSelect with empty validation object', () => {
  let props = {
    options, onChange,
    validation: {}
  }

  const component = renderer.create(<AmfSelect {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfSelect test onChange function', () => {
  let props = {
    options, onChange
  }

  const component = new AmfSelect(props)
  component.onChange('foo', 0)
  
  expect(props.onChange).toHaveBeenCalled()
})

test('AmfSelect, test willHide event', () => {
  let props = {
    options, onChange
  }

  const component = new AmfSelect(props)
  component.setState = jest.fn()
  component.onSelectWillHide()
  
  expect(component.setState).toHaveBeenCalled()
})

test('AmfSelect, test willShow event', () => {
  let props = {
    options, onChange
  }
  let propsWithValidation = {
    ...props, validation, value: null
  }

  const component = new AmfSelect(props)
  component.setState = jest.fn()
  component.onSelectWillShow()

  const componentWithValidation = new AmfSelect(propsWithValidation)
  componentWithValidation.setState = jest.fn()
  componentWithValidation.onSelectWillShow()

  expect(component.setState).toHaveBeenCalled()
  expect(propsWithValidation.onChange).toHaveBeenCalled()
})

test('AmfSelect, test didUpdate event', () => {
  let props = {
    options, onChange
  }

  const component = new AmfSelect(props)
  component.state = {isFocused: false}
  component.validate = jest.fn()
  component.componentDidUpdate({}, {isFocused: true})
  expect(component.validate).toHaveBeenCalled()

  const componentFocused = new AmfSelect(props)
  componentFocused.state = {isFocused: true}
  componentFocused.validate = jest.fn()
  componentFocused.componentDidUpdate({}, {isFocused: false})
  expect(componentFocused.validate).not.toHaveBeenCalled()
})

test.skip('AmfSelect, test new options', () => {
  let props = {
    options, onChange
  }
  let newItem = [{
    label: 'new',
    value: 'new'
  }]
  let expected = newItem.map(i => i.label)

  const component = new AmfSelect(props)
  component.componentWillReceiveProps({options: newItem})

  expect(component.options).toEqual(expected)
})
