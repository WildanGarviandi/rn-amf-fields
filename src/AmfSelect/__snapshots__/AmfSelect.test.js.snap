// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmfSelect with custom selector 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect with empty validation object 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect with empty value 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "red",
        "margin": 4,
        "marginBottom": 0,
      }
    }
  >
    undefined harus dipilih
  </Text>
</View>
`;

exports[`AmfSelect with horizontal layout 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect with label 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": "Label",
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect with label 2`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": "Label",
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect with required validation 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect with required validation 2`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={1}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect with unmatch value 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;

exports[`AmfSelect without validation 1`] = `
<View>
  <View
    style={undefined}
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": undefined,
            "textColor": undefined,
            "value": undefined,
          },
          Object {
            "label": "foo",
            "textColor": undefined,
            "value": "foo",
          },
          Object {
            "label": "bar",
            "textColor": undefined,
            "value": "bar",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#4b6ac5",
        "borderColor": "#4b6ac5",
        "borderStyle": "solid",
        "height": 3,
      }
    }
  />
</View>
`;
